[com4]: output for release area information

PaulaSp3 opened the pull request:
- rasters are only merged if they are written as output.

- computation time (for my example): +25%
- new outputs:
  - affected area for a segmented release area (that is marked by a release ID) (polygon)
  - Rel Id count (how many release areas - marked with a release ID) hit a cell (raster)
- outputs that were new within this PR, but are deleted with this commit [b8be3c9](https://github.com/OpenNHM/AvaFrame/pull/1213/changes/b8be3c9b8606749e4e4fe1e741a8fd1af63e1266) again and could be restored with this commit:
  - mimimum volume of release area (raster)
  - maximum volume of release area (raster)

- required input data (for these outputs):
  - the release ID is given as a raster (the release ID is the raster value). Cells with the same release ID belong the same release area. The raster is provided in Inputs/RELID.

- added tests for various functions for tiling and merging the tiles

PaulaSp3 committed:
compute startCellID per cell and compute output polygon for each process path

still in progress

relVolMin and relVolMax as output

correct position for relId

delete print

small change

fix bug

pathPolygons as output

rename function

tidy up

try to keep RAM as small as possible

print a warning if the output in cfg is not completly valid

small corrections

correct bugs and doc

update docu

adapt pytest

delete unused package

Update com4FlowPy.py deleteTempFolder - small BugFix and adaptation for releaseID functionality

add tests for tiling, merging of tiles

fixed bug

extend split and merge test

add test for dictToPolygon

fix bug

format and test fix

ahuber-bfw commented on docs/moduleCom4FlowPy.rst:
```diff
@@ -244,6 +246,10 @@ In addition these output layers are also available:
 - ``depFluxSum``: deposited flux summed up over all paths
 - ``travelLengthMin``: the travel length along the flow path (the minimum value of all paths for every raster cell)
 - ``fpTravelAngleMin``: the gamma angle along the flow path (the minimum value of all paths for every raster cell)
+- ``relVolMin``: the minimum of the tracked release volume that route through the raster cell (the volume is provided in the release file in the REL folder)
```
generell die Frage, wie man mit den extra input-files umgehen sollte, damit die Vorgehensweise konsistent bleibt (auch im Hinblick darauf, wie man es macht, falls in Zukunft noch zusätzliche inputs dazu kommen):

 akutell:
-  releaseId --> aktuell eigener releaseIDPath oder .tif in RELID\ ABER
-  releaseVolume --> implizit in \REL bzw. releasePath

konsistent(er)?:
- releaseID und releaseVolume als extra layer in \REL mit fixer Vorgabe für fileName bzw. jeweils mit eigenem releaseIdPath und releaseVolumePath ODER
- \REL \RELID \RELVOLUME als eigene folder in inputs bzw. releasePath, releaseIdPath und releaseVolumePath

ahuber-bfw commented on avaframe/runCom4FlowPy.py:
```diff
@@ -291,6 +292,20 @@ def readFlowPyinputs(avalancheDir, cfgFlowPy, log):
         forestPath = ""
     cfgPath["forestPath"] = forestPath
 
+    # read release ID raster
```
Gleicher kommentar, wie in der Doku:
* Macht's Sinn auch den releaseVolume input in ein extra file zu geben anstatt implizit im release file?
* Wenn ja, dann würd hier auch noch ein entsprechender Input-Check sinnvoll sein

ahuber-bfw commented on avaframe/com4FlowPy/splitAndMerge.py:
```diff
@@ -233,3 +236,117 @@ def mergeRaster(inDirPath, fName, method='max'):
     return mergedRas
+
+
+def mergeDict(inDirPath, fName):
```
maybe consider writing a test for the new functions in `tests/test_com4FlowPy.py`

ahuber-bfw commented on avaframe/com4FlowPy/flowClass.py:
```diff
@@ -177,6 +191,10 @@ def add_parent(self, parent):
             if parent.forestIntCount < (self.forestIntCount - self.isForest):
                 self.forestIntCount = parent.forestIntCount + self.isForest
 
+    def calc_startCellVol(self, startcellVolNew):
```
generally - what is the use-case of tracking `volMin` and `volMax`?

- is it size-classification on the merged model results?
- is there a use-case, where `volMin` and `volMax` are interesting and you don't use the `relID` option at the same time?

overall the naming von `volMin` and `volMax` could be a bit misleading/confusing - mainly, because it is physiclly improbable, that the whole release Volume is routed through each cell of a modeled process zone (which maybe could be represented be some combination with the routing flux?). 

If this functionality is always/only interesting in conjunction with also using `relIDs` for contiguous release areas, could you simply add the information about min and max volumes at the final merging step in `com4FlowPy.py`?

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -196,6 +199,16 @@ def run(optTuple):
+    if relVolBool:
+        relVolArray = release.copy()
+    else:
+        relVolArray = None
```
also here:

* does it make sense to provide release volume in a separate file and not implicitly in the release file?

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -275,6 +295,9 @@ def run(optTuple):
         forestIntArray = np.ones_like(dem, dtype=np.float32) * -9999
+    relVolMinArray = np.ones_like(dem, dtype=np.float32) * -9999
+    relVolMaxArray = np.zeros_like(dem, dtype=np.float32)
+    processedStartCellIdDict = {}
```
only initialize these if the corresponding options in `relOutputParams` are set to `True`?

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -289,6 +312,9 @@ def run(optTuple):
     travelLengthMaxList = []
     travelLengthMinList = []
+    processedStartCellIdList = []
+    relVolMinList = []
```
same here

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -308,8 +334,11 @@ def run(optTuple):
         if forestInteraction:
-            forestIntList.append(res[12])
+            forestIntList.append(res[15])
+        relVolMinList.append(res[13])
```
and here

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -482,6 +537,9 @@ def updateInfraDirGraph(row, col, parentRow=None, parentCol=None):
     travelLengthMinArray = np.ones_like(dem, dtype=np.float32) * -9999
     travelLengthMaxArray = np.ones_like(dem, dtype=np.float32) * -9999
 
+    relVolMinArray = np.ones_like(dem, dtype=np.float32) * -9999
```
only initialize if option is set, else initialize with `None`?
in this way we do not reserve any memory if the array is not used in the next steps + also the return statement (`lines 813 - 830`) at the end of the function could stay like it is (all possible results are returned, but the ones which are not calculated are set to `None` - higher level functions can then also easily check, if a give result is available ....

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -543,10 +613,12 @@ def updateInfraDirGraph(row, col, parentRow=None, parentCol=None):
             fluxDistOldVersionBool=fluxDistOldVersionBool,
             FSI=forestArray[row_idx, col_idx] if isinstance(forestArray, np.ndarray) else None,
             forestParams=forestParams,
+            startcellVol=startcellVol,
```
see comments in `flowClass.py`

--> is it really necessary to implement this at the level of single cells (since it only passes one particular information from release cell to all the cells in the process area, if I understand correctly?) 
- maybe this can be accomplished at a higher level or in a quick post-processing step?

ahuber-bfw reviewed (changes requested):
Hey @PaulaSp3,

Cool features (esp. the releaseID to Polygon feature :) )

only some questions/remarks:

- maybe you can add some tests for the new functions in `splitAndMerge.py` in the `test_com4FlowPy.py` file 
- handling of additional output arrays in `flowCore.py` --> to be consistent with the method used for other optional outupts (e.g. infra, forestInteraction) maybe we could only initiate the required arrays, if the option is set in the .ini - else just assign `None`?
- regarding the option to have a `relVolMin` and `relVolMax` output
    - are there use-cases, where this is interesting and you don't also use the relID option at the same time?
    - if this is only used togehter with the relID option, then this could probably be implemented at a higher level or in a post-processing step without need for modification in `flowCore.py` and `flowClass.py`

ahuber-bfw commented on avaframe/runCom4FlowPy.py:
```diff
@@ -163,6 +163,7 @@ def main(avalancheDir=''):
         cfgPath["tempDir"] = pathlib.Path(temp_dir)
         cfgPath["demPath"] = pathlib.Path(cfgCustomPaths["demPath"])
         cfgPath["releasePath"] = pathlib.Path(cfgCustomPaths["releasePath"])
+        cfgPath["relIdPath"] = pathlib.Path(cfgCustomPaths["relIdPath"])
```
im `com4FlowPyCfg.ini` heißt die variable `releaseIdPath` --> wird deshalb hier so nicht gefunden

ahuber-bfw commented on avaframe/com4FlowPy/com4FlowPyCfg.ini:
```diff
@@ -256,6 +262,7 @@ useCustomPathDEM = False
 workDir =
 demPath =
 releasePath =
+releaseIdPath =
```
im `runCom4FlowPy.py` wird nach `relIdPath` gesucht --> hier oder dort anpassen damits funktioniert

PaulaSp3 committed:
delete release volume min and max output option

comments AH

Update avaframe/tests/test_com4FlowPy.py

Co-authored-by: qltysh[bot] <168846912+qltysh[bot]@users.noreply.github.com>

fix bug

clean and make polygon geometry valid

delete temp files

delete unued import

comments AH

Update avaframe/tests/test_com4FlowPy.py

Co-authored-by: ahuber-bfw <112931921+ahuber-bfw@users.noreply.github.com>

comment AH

ahuber-bfw commented on avaframe/com4FlowPy/flowCore.py:
```diff
@@ -474,32 +512,44 @@ def updateInfraDirGraph(row, col, parentRow=None, parentCol=None):
     row_list, col_list = get_start_idx(dem, release)
 
     startcell_idx = 0
+    startCellIdDict = {}
```
auch nur nötig `if relIdBool`?
wenn ja, dann vlt auch explizit hinschreiben, dann ist gleich klar, was die Zeile macht und es wird nicht per default ein leeres dict initialisiert

ahuber-bfw commented on avaframe/tests/test_com4FlowPy.py:
```diff
@@ -194,32 +169,354 @@ def test_calculation():
+
+
+def test_mergeDictToPolygon():
+    pathTestFolder = pathlib.Path("avaframe/tests/data/testCom4")
```
currently throws a "file not found" error, when `pytest` is not invoked in `AvaFrame` directory - this way it should work from anywhere: 
```suggestion
    _testDIR = pathlib.Path(__file__).parent 
    pathTestFolder = _testDIR /  "data" / "testCom4"
```

ahuber-bfw commented on avaframe/com4FlowPy/splitAndMerge.py:
```diff
@@ -207,7 +209,7 @@ def mergeRaster(inDirPath, fName, method='max'):
 
     mergedRas = np.zeros((extL[0], extL[1]))
```
maybe we can be more explicit here - sth. like:

```suggestion
    if method in ["max", "min"]:
        mergedRas = np.full((extL[0], extL[1]), np.nan)
    elif method == "sum":
        mergedRas = np.zeros((extL[0], extL[1]))
    else:
        raise ValueError(
            f"Invalid function argument <method = '{method}'>. Supported values for 'method': ['min', 'max', 'sum']"
        )
```

ahuber-bfw commented on avaframe/tests/data/testCom4/refPolygon_manipulated.qmd:
```diff
@@ -0,0 +1,27 @@
+<!DOCTYPE qgis PUBLIC 'http://mrcc.com/qgis.dtd' 'SYSTEM'>
```
is the `.qmd` file needed for something? - if not I would delete it / not track it

ahuber-bfw reviewed (approved):
looks good!
only some minor comments left - see below.
